c++/src/kj/compat/http.c++

```diff
@@ -4345,16 +4753,25 @@ public:
 
     kj::Maybe<kj::String> offeredExtensions;
     kj::Maybe<CompressionParameters> clientOffer;
-    if (settings.enableWebSocketCompression) {
+    kj::Vector<CompressionParameters> extensions;
+    auto compressionMode = settings.webSocketCompressionMode;
+
+    if (compressionMode == HttpClientSettings::MANUAL_COMPRESSION) {
       KJ_IF_MAYBE(value, headers.get(HttpHeaderId::SEC_WEBSOCKET_EXTENSIONS)) {
         // Strip all `Sec-WebSocket-Extensions` except for `permessage-deflate`.
-        auto extensions = _::findValidExtensionOffers(*value);
-        if (extensions.size() > 0) {
-          clientOffer = extensions.front();
-          connectionHeaders[HttpHeaders::BuiltinIndices::SEC_WEBSOCKET_EXTENSIONS] =
-              offeredExtensions.emplace(_::generateExtensionRequest(extensions.asPtr()));
-        }
+        extensions = _::findValidExtensionOffers(*value);
       }
+    } else if (compressionMode == HttpClientSettings::AUTOMATIC_COMPRESSION) {
+      // If AUTOMATIC_COMPRESSION is enabled, we send `Sec-WebSocket-Extensions: permessage-deflate`
+      // to the server and ignore the `headers` provided by the caller.
+      constexpr auto default_config = "permessage-deflate";
+      extensions = _::findValidExtensionOffers(default_config);
+    }
+
+    if (extensions.size() > 0) {
+      clientOffer = extensions.front();
```

Comment:
I still don't really understand why we allow multiple extension offers to be sent, but then primarily care about the first one. Sorry if I asked this already... what's the use case for sending multiple offers? And why do we specifically remember the first offer we sent and not the others?

Reply:
> Why do we allow multiple offers to be sent?

Right, so it really boils down to the fact that the server might reject the client's requested parameters. If we don't have fallback options, the connection won't use compression at all.

Let's say we're the client and we send `permessage-deflate; server_max_window_bits = 10`. The server gets this and doesn't recognize `server_max_window_bits`, so it rejects the offer and replies without the `Sec-WebSocket-Extensions` header (i.e. we establish a regular websocket connection). Here's an example from the rfc (https://datatracker.ietf.org/doc/html/rfc7692#section-7.1.3)

```
        Sec-WebSocket-Extensions:
           permessage-deflate;
           client_max_window_bits; server_max_window_bits=10

   This extension negotiation offer might be rejected by the server
   because the server doesn't support the "server_max_window_bits"
   extension parameter in an extension negotiation offer.  This is fine
   if the client cannot receive messages compressed using a larger
   sliding window size, but if the client just prefers using a small
   window but wants to fall back to the "permessage-deflate" without the
   "server_max_window_bits" extension parameter, the client can make an
   offer with the fallback option like this:

       Sec-WebSocket-Extensions:
           permessage-deflate;
           client_max_window_bits; server_max_window_bits=10,
           permessage-deflate;
           client_max_window_bits

   The server can accept "permessage-deflate" by picking any supported
   one from the listed offers.  To accept the first option, for example,
   the server may send back a response as follows:

       Sec-WebSocket-Extensions:
           permessage-deflate; server_max_window_bits=10

   To accept the second option, for example, the server may send back a
   response as follows:

       Sec-WebSocket-Extensions: permessage-deflate
```
---

> Why do we remember the first offer we sent and not the others?

Couple things to note here:
1. Offers are ordered by preference.
2. We can assume that offer `n` has more parameters than offer `n + 1`.
3. We only want to retain `client_no_context_takeover` and `client_max_window_bits` (the client's compressor config).

So first, why retain this info at all? Basically, the `client_x` parameters are hints to the server about how the client will configure its _compressor_. The server isn't required to acknowledge these hints (by replying with the same parameters in the response), so if we have a situation where the client sends `client_no_context_takeover` and the server ignores it, then the only way the client can use `client_no_context_takeover` is if it remembered that it sent it in the first place.

As for why we only retain the first one, see (1) and (2). The assumption is that if a client wants to use any `client_x` parameters, then offer 1 will have all the `client_x` params that they'd want to use.

Reply:
OK, I get it now. Thanks.

Maybe worthwhile to explain this in a code comment.

Reply:
https://github.com/capnproto/capnproto/pull/1505/commits/c350aaf975f89d5e16aaa613c3535ce5d21112d1